### PR TITLE
getQuery() returns all when null is passed

### LINKED ISSUE #10256: Request::getQuery() cannot get full query.
* [ ] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3.4.0
* Platform and Target: Apache2, MySQL 5,7, PHP 7.1.1

### What you did

In 3.3 you could get the entire request query.

```
   var_dump($this->request->query());
```

### What happened

Doing the same in 3.4 yields `null`

```
   var_dump($this->request->getQuery());
```

Looks like there was some oversight on updating to getters and setters.

- ServerRequest::query() is deprecated, but it did not follow the pattern of duel getter/setter arguments for the other deprecated methods (i.e. you can not set values with this method)
- ServerRequest::query(null) returns the enture query.
- ServerRequest::getQuery() does not support null parameter.
- some deprecated messages refer to setQuery and withQuery, but those methods do not exist.

If this RFC is valid, then let me know and I'll open a PR to fix the above.

### PR DESCRIPTION
**Pull request by thinkingmedia**
Closes #10256

### CODE REVIEW

**Comment by thinkingmedia** [comment]
Sorry, I keep writing `getData` in the comments, but it really is `getQuery`

File: src/Http/ServerRequest.php [hunk 1]
```diff
@@ -1410,12 +1422,26 @@ public function query($name = null)
     /**
      * Read a specific query value or dotted path.
      *
-     * @param string $name The name or dotted path to the query param.
-     * @param mixed $default The default value if the named parameter is not set.
+     * Developers are encouraged to use getQueryParams() when possible as it is PSR-7 compliant, and this method
+     * is not. You can future proof your code by using the following example.
```
Comment (dereuromark) [inline src/Http/ServerRequest.php]: Are we sure we want to get rid of the Hash::get() convenience method here in future versions?
To me it sounds like a good idea to keep `getQuery('Post.id')` as an easier way of `Hash::get($request->getQueryParams(), 'Post.id', null);`.

Comment (thinkingmedia) [inline src/Http/ServerRequest.php]: @dereuromark If you want to use Guzzle as your middleware then getQuery is not a PSR-7 method, and won't exist in the request object. The point of the comment is to advise developers that if they want to be PSR-7 compliant, then don't use getQuery. It's just that ServerRequestInterface is the lower common interface between Guzzle and Cake.

I'm not saying the method will go away. But if you write your code as in the example, then you're code is more PSR-7 compliant. 

I think this was the whole point of PSR-7 support in 3.4, but I'm not completely sure on how it's going to work. I don't think we're completely there yet (i.e. you can't use Guzzle as the middleware).

Maybe I'm wrong about all this. I don't know.

Comment (ADmad) [inline src/Http/ServerRequest.php]: > you can't use Guzzle as the middleware

Well Guzzle itself is not a middleware, it's a HTTP client lib which is PSR-7 compliant.

Comment (ADmad) [inline src/Http/ServerRequest.php]: The argument you are probably looking for it if one makes a middleware which uses `$request->getQuery()` then that middleware would only be usable with CakePHP.

Comment (thinkingmedia) [inline src/Http/ServerRequest.php]: @ADmad okay, I have no idea how all these PSR-7 interfaces improve Cake. If any of the comments I added to the code are invalid. Please let me know, and I'll change it.

Comment (ADmad) [inline src/Http/ServerRequest.php]: Just remove / modify the statement `You can future proof your code..` as it be mistaken to mean that this method will be removed in future.

**Review by markstory — CHANGES_REQUESTED** [review]
Looks good, but it needs a test for the new return value/argument pair.

File: src/Http/ServerRequest.php [hunk 2]
```diff
@@ -1006,6 +1006,7 @@ public function getHeaders()
      *
      * @param string $name The header you want to get (case-insensitive)
      * @return bool Whether or not the header is defined.
+     * @link http://www.php-fig.org/psr/psr-7/ This method is part of the PSR-7 server request interface.
```
Comment (ravage84) [inline src/Http/ServerRequest.php]: 👍

**Comment by thinkingmedia** [comment]
My stupid rebase failed.

**Comment by thinkingmedia** [comment]
This PR was moved here: https://github.com/cakephp/cakephp/pull/10278

**Pull request closed without merging**